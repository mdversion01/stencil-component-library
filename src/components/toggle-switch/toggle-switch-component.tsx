// src/components/toggle-switch/toggle-switch-component.tsx
import { Component, Prop, State, Event, EventEmitter, h, Watch } from '@stencil/core';
import type { ToggleItem } from './toggle-switch-types';

@Component({
  tag: 'toggle-switch-component',
  styleUrls: ['../form-styles.scss', 'toggle-switch.scss'],
  shadow: false,
})
export class ToggleSwitchComponent {
  @Prop() inputId: string = '';
  @Prop() checked: boolean = false;
  @Prop() disabled: boolean = false;
  @Prop() required: boolean = false;
  @Prop() labelTxt: string = '';
  @Prop() size: string = '';
  @Prop() inline: boolean = false;
  @Prop() value: string = '';
  @Prop() toggleTxt: boolean = false;
  @Prop() newToggleTxt: { on: string; off: string } = { on: 'On', off: 'Off' };
  @Prop() validation: boolean = false;
  @Prop() validationMessage: string = '';

  @Prop() switches: boolean = false;
  @Prop() switchesArray: ToggleItem[] = [];

  // NEW: choose custom vs bootstrap
  @Prop() customSwitch: boolean = false;

  @State() keyboardFocused: boolean = false;
  @State() isChecked: boolean = false;
  @State() internalSwitchesArray: ToggleItem[] = [];

  @Event({ eventName: 'checkedChanged', bubbles: true, composed: true })
  checkedChanged: EventEmitter<{ id: string; checked: boolean }>;

  componentWillLoad() {
    this.isChecked = this.checked;
    this.internalSwitchesArray = [...this.switchesArray];
  }

  @Watch('switchesArray')
  syncSwitchesArray(newValue: ToggleItem[]) {
    this.internalSwitchesArray = [...newValue];
  }

  @Watch('checked')
  handleCheckedChange(newValue: boolean) {
    this.isChecked = newValue;
  }

  componentDidLoad() {
    document.addEventListener('keydown', this.trackKeyboard);
    document.addEventListener('mousedown', this.trackMouse);
    document.addEventListener('touchstart', this.trackMouse);
  }

  disconnectedCallback() {
    document.removeEventListener('keydown', this.trackKeyboard);
    document.removeEventListener('mousedown', this.trackMouse);
    document.removeEventListener('touchstart', this.trackMouse);
  }

  private trackKeyboard = (e: KeyboardEvent) => {
    if (e.key === 'Tab') this.keyboardFocused = true;
  };
  private trackMouse = () => {
    this.keyboardFocused = false;
  };

  private toggleSwitch(index: number) {
    const item = this.internalSwitchesArray[index];
    if (this.disabled || item.disabled) return;

    const updated = [...this.internalSwitchesArray];
    updated[index].checked = !updated[index].checked;
    this.internalSwitchesArray = updated;
    this.checkedChanged.emit({ id: updated[index].id, checked: updated[index].checked });
  }

  private toggleSingle = () => {
    if (this.disabled) return;
    this.isChecked = !this.isChecked;
    this.checkedChanged.emit({ id: this.inputId, checked: this.isChecked });
  };

  // ---------- RENDER HELPERS (MULTI) ----------
  private renderBootstrapSwitch(item: ToggleItem, parentId: string, index: number) {
    const switchId = `${parentId}_option_${item.id}`;
    const invalid = item.required && item.validation && !item.checked;
    const showToggleTxt = (item.toggleTxt ?? this.toggleTxt) === true;
    const toggleText = item.checked ? (item.newToggleTxt?.on || this.newToggleTxt.on) : (item.newToggleTxt?.off || this.newToggleTxt.off);

    // size: bootstrap doesn't have size variants for switches by default; keep your size hook as modifier classes if you style them
    const wrapperCls = [
      'form-check',
      'form-switch',
      this.inline ? 'form-check-inline' : '',
      this.size ? `ts-size-${item.size || this.size}` : '', // optional hook for custom sizing
    ].filter(Boolean).join(' ');

    return (
      <div class={wrapperCls}>
        <input
          id={switchId}
          class={`form-check-input ${invalid ? 'is-invalid' : ''}`}
          type="checkbox"
          role="switch"
          checked={item.checked}
          disabled={this.disabled || item.disabled}
          required={this.required || item.required}
          aria-checked={String(item.checked)}
          aria-disabled={String(this.disabled || item.disabled)}
          onChange={() => this.toggleSwitch(index)}
          value={item.value}
        />
        <label
          class="form-check-label"
          htmlFor={switchId}
          onMouseDown={e => e.preventDefault()}
        >
          {item.label}
          {showToggleTxt && <span class="toggleTxt-bold"> {toggleText}</span>}
          {item.required && <span class="required">*</span>}
        </label>

        {!this.inline && invalid && item.validationMessage && (
          <div class="invalid-feedback">{item.validationMessage}</div>
        )}
      </div>
    );
  }

  private renderCustomSwitch(item: ToggleItem, parentId: string, index: number) {
    const switchId = `${parentId}_option_${item.id}`;
    const toggleText = item.checked ? item.newToggleTxt?.on || this.newToggleTxt.on : item.newToggleTxt?.off || this.newToggleTxt.off;
    const showToggleTxt = item.toggleTxt ?? this.toggleTxt;
    const invalid = item.required && item.validation && !item.checked;

    if (this.inline) {
      return (
        <div class={`custom-control custom-switch ${item.size || this.size ? `custom-control-${item.size || this.size}` : ''} custom-control-inline`}>
          <div>
            <input
              id={switchId}
              type="checkbox"
              class="custom-control-input"
              role="switch"
              checked={item.checked}
              disabled={this.disabled || item.disabled}
              required={this.required || item.required}
              aria-checked={String(item.checked)}
              aria-disabled={String(this.disabled || item.disabled)}
              onChange={() => this.toggleSwitch(index)}
              value={item.value}
            />
            <label
              htmlFor={switchId}
              onMouseDown={e => e.preventDefault()}
              class={`custom-control-label ${this.keyboardFocused ? 'keyboard-focused' : ''} ${invalid ? 'invalid' : ''}`}
            >
              {item.label}
              {showToggleTxt && <span class="toggleTxt-bold"> {toggleText}</span>}
              {item.required && <span class="required">*</span>}
            </label>
          </div>
        </div>
      );
    }

    return (
      <div class={`custom-control custom-switch ${item.size || this.size ? `custom-control-${item.size || this.size}` : ''}`}>
        <input
          id={switchId}
          type="checkbox"
          class="custom-control-input"
          role="switch"
          checked={item.checked}
          disabled={this.disabled || item.disabled}
          required={this.required || item.required}
          aria-checked={String(item.checked)}
          aria-disabled={String(this.disabled || item.disabled)}
          onChange={() => this.toggleSwitch(index)}
          value={item.value}
        />
        <label
          htmlFor={switchId}
          onMouseDown={e => e.preventDefault()}
          class={`custom-control-label ${this.keyboardFocused ? 'keyboard-focused' : ''} ${invalid ? 'invalid' : ''}`}
        >
          {item.label}
          {showToggleTxt && <span class="toggleTxt-bold"> {toggleText}</span>}
          {item.required && <span class="required">*</span>}
        </label>

        {invalid && item.validationMessage && <div class="invalid-feedback">{item.validationMessage}</div>}
      </div>
    );
  }

  private renderSwitch(item: ToggleItem, parentId: string, index: number) {
    return this.customSwitch
      ? this.renderCustomSwitch(item, parentId, index)
      : this.renderBootstrapSwitch(item, parentId, index);
  }

  // ---------- RENDER (SINGLE) ----------
  private renderSingleBootstrap() {
    const switchId = this.inputId || 'toggle-switch-' + Math.random().toString(36).substr(2, 5);
    const toggleText = this.isChecked ? this.newToggleTxt.on : this.newToggleTxt.off;
    const invalid = this.required && this.validation && !this.isChecked;

    const wrapperCls = [
      'form-check',
      'form-switch',
      this.inline ? 'form-check-inline' : '',
      this.size ? `ts-size-${this.size}` : '',
    ].filter(Boolean).join(' ');

    return (
      <div class="form-group toggle-switch">
        <div class={wrapperCls}>
          <input
            id={switchId}
            class={`form-check-input ${invalid ? 'is-invalid' : ''}`}
            type="checkbox"
            role="switch"
            checked={this.isChecked}
            disabled={this.disabled}
            required={this.required}
            aria-checked={String(this.isChecked)}
            aria-disabled={String(this.disabled)}
            onChange={this.toggleSingle}
            value={this.value}
          />
          <label
            class="form-check-label"
            htmlFor={switchId}
            onMouseDown={e => e.preventDefault()}
          >
            {this.labelTxt}
            {this.toggleTxt && <span class="toggleTxt-bold"> {toggleText}</span>}
            {this.required && <span class="required">*</span>}
          </label>

          {invalid && this.validationMessage && <div class="invalid-feedback">{this.validationMessage}</div>}
        </div>
      </div>
    );
  }

  private renderSingleCustom() {
    const switchId = this.inputId || 'toggle-switch-' + Math.random().toString(36).substr(2, 5);
    const toggleText = this.isChecked ? this.newToggleTxt.on : this.newToggleTxt.off;
    const invalid = this.required && this.validation && !this.isChecked;

    return (
      <div class="form-group toggle-switch">
        <div class={`custom-control custom-switch ${this.size ? `custom-control-${this.size}` : ''} ${this.inline ? 'custom-control-inline' : ''}`}>
          <input
            id={switchId}
            class="custom-control-input"
            type="checkbox"
            role="switch"
            checked={this.isChecked}
            disabled={this.disabled}
            required={this.required}
            aria-checked={String(this.isChecked)}
            aria-disabled={String(this.disabled)}
            onChange={this.toggleSingle}
            value={this.value}
          />
          <label
            htmlFor={switchId}
            onMouseDown={e => e.preventDefault()}
            class={`custom-control-label ${this.keyboardFocused ? 'keyboard-focused' : ''} ${invalid ? 'invalid' : ''}`}
          >
            {this.labelTxt}
            {this.toggleTxt && <span class="toggleTxt-bold"> {toggleText}</span>}
            {this.required && <span class="required">*</span>}
          </label>
          {invalid && this.validationMessage && <div class="invalid-feedback">{this.validationMessage}</div>}
        </div>
      </div>
    );
  }

  render() {
    const parentId = this.inputId || 'toggle-group';

    if (this.switches) {
      const invalidItems = this.internalSwitchesArray.filter(i => i.required && i.validation && !i.checked);
      const invalidMessages = invalidItems.map(i => i.validationMessage?.trim()).filter(Boolean) as string[];
      const hasAnyInvalid = invalidMessages.length > 0;

      return (
        <div class="form-group ts-group">
          <div
            id={parentId}
            role="group"
            class={this.inline ? (this.customSwitch ? '' : 'form-toggle-inline') : ''} // keep your class hook if needed
            aria-invalid={hasAnyInvalid ? 'true' : undefined}
            aria-describedby={hasAnyInvalid ? `${parentId}-validation` : undefined}
          >
            {this.internalSwitchesArray.map((item, i) => this.renderSwitch(item, this.inputId || parentId, i))}
          </div>

          {this.inline && hasAnyInvalid && (
            <div id={`${parentId}-validation`} class="ts-inline invalid-feedback">
              {invalidMessages.map((msg, idx) => (
                <div key={idx}>{msg}</div>
              ))}
            </div>
          )}
        </div>
      );
    }

    // single
    return this.customSwitch ? this.renderSingleCustom() : this.renderSingleBootstrap();
  }
}
