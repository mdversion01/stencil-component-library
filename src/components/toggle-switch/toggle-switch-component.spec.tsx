// src/components/toggle-switch/toggle-switch-component.spec.tsx

import { newSpecPage } from '@stencil/core/testing';
import { ToggleSwitchComponent } from './toggle-switch-component';

const createWithSwitchesArray = async switchesArray => {
  const page = await newSpecPage({
    components: [ToggleSwitchComponent],
    html: `<toggle-switch-component switches="true"></toggle-switch-component>`,
  });
  page.rootInstance.syncSwitchesArray(switchesArray);
  await page.waitForChanges();
  return page;
};

describe('toggle-switch-component', () => {
  it('renders a toggle with toggleTxt and newToggleTxt', async () => {
    const page = await newSpecPage({
      components: [ToggleSwitchComponent],
      html: `<toggle-switch-component
               input-id="txt-toggle"
               label-txt="Power"
               toggle-txt="true"
               checked="true"
               new-toggle-txt='{ "on": "Enabled", "off": "Disabled" }'>
             </toggle-switch-component>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders multiple switches from switchesArray', async () => {
    const page = await createWithSwitchesArray([
      { id: 'one', label: 'First', checked: false },
      { id: 'two', label: 'Second', checked: true },
    ]);
    expect(page.root.querySelectorAll('input[type="checkbox"]').length).toBe(2);
    expect(page.root).toMatchSnapshot();
  });

  it('emits checkedChanged on single toggle', async () => {
    const page = await newSpecPage({
      components: [ToggleSwitchComponent],
      html: `<toggle-switch-component input-id="single-toggle" label-txt="Toggle Me"></toggle-switch-component>`,
    });

    const spy = jest.fn();
    page.win.addEventListener('checkedChanged', spy);

    const input = page.root.querySelector('input[type="checkbox"]');
    input.dispatchEvent(new Event('change'));
    await page.waitForChanges();

    expect(spy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: { id: 'single-toggle', checked: true },
      }),
    );
  });

  it('applies correct aria-* attributes', async () => {
    const page = await newSpecPage({
      components: [ToggleSwitchComponent],
      html: `<toggle-switch-component input-id="aria-toggle" checked="false" disabled="true"></toggle-switch-component>`,
    });

    const input = page.root.querySelector('input');
    expect(input.getAttribute('aria-checked')).toBe('false');
    expect(input.getAttribute('aria-disabled')).toBe('true');
  });

  it('shows validation error when required and invalid', async () => {
    const page = await newSpecPage({
      components: [ToggleSwitchComponent],
      html: `<toggle-switch-component input-id="val-toggle" required="true" validation="true" label-txt="Terms" validation-message="Required field."></toggle-switch-component>`,
    });

    const error = page.root.querySelector('.invalid-feedback');
    expect(error).toBeTruthy();
    expect(error.textContent.trim()).toBe('Required field.');
  });

  it('sets keyboardFocused true on Tab press', async () => {
    const page = await newSpecPage({
      components: [ToggleSwitchComponent],
      html: `<toggle-switch-component input-id="tab-toggle"></toggle-switch-component>`,
    });

    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Tab' }));
    await page.waitForChanges();

    expect(page.rootInstance.keyboardFocused).toBe(true);
  });

  it('toggles switchesArray item and emits event', async () => {
    const page = await createWithSwitchesArray([
      { id: 'item1', label: 'Wi-Fi', checked: false },
      { id: 'item2', label: 'Bluetooth', checked: true },
    ]);

    const spy = jest.fn();
    page.win.addEventListener('checkedChanged', spy);

    const inputs = page.root.querySelectorAll('input[type="checkbox"]');
    inputs[0].dispatchEvent(new Event('change'));
    await page.waitForChanges();

    expect(spy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: { id: 'item1', checked: true },
      }),
    );
  });

  it('renders inline and size class combinations (default BS5 switch)', async () => {
    const page = await newSpecPage({
      components: [ToggleSwitchComponent],
      html: `<toggle-switch-component
             input-id="styled-toggle"
             label-txt="Wi-Fi"
             size="lg"
             inline="true">
           </toggle-switch-component>`,
    });

    // Default = Bootstrap 5 switch
    const wrapper = page.root.querySelector('.form-check.form-switch') as HTMLElement;
    expect(wrapper).toBeTruthy();

    // Inline uses BS5 class
    expect(wrapper.classList.contains('form-check-inline')).toBe(true);

    // BS5 doesn’t ship a built-in "lg" size class for switches.
    // If you didn’t add a custom size mapping for the default switch,
    // just assert that the old custom-control size class is NOT present.
    expect(wrapper.classList.contains('custom-control-lg')).toBe(false);
  });

  it('renders inline and size class combinations when customSwitch is true', async () => {
    const page = await newSpecPage({
      components: [ToggleSwitchComponent],
      html: `<toggle-switch-component
             custom-switch="true"
             input-id="styled-toggle"
             label-txt="Wi-Fi"
             size="lg"
             inline="true">
           </toggle-switch-component>`,
    });

    // Custom = your legacy/custom classes
    const wrapper = page.root.querySelector('.custom-control.custom-switch') as HTMLElement;
    expect(wrapper).toBeTruthy();
    expect(wrapper.classList.contains('custom-control-lg')).toBe(true);
    expect(wrapper.classList.contains('custom-control-inline')).toBe(true);
  });

  it('emits events for multiple switches toggled in quick succession', async () => {
    const page = await createWithSwitchesArray([
      { id: 'wifi', label: 'Wi-Fi', checked: false },
      { id: 'bt', label: 'Bluetooth', checked: false },
      { id: 'gps', label: 'GPS', checked: false },
    ]);

    const spy = jest.fn();
    page.win.addEventListener('checkedChanged', spy);

    const inputs = page.root.querySelectorAll('input[type="checkbox"]');
    inputs.forEach(input => input.dispatchEvent(new Event('change')));
    await page.waitForChanges();

    expect(spy).toHaveBeenCalledTimes(3);
    expect(spy).toHaveBeenNthCalledWith(1, expect.objectContaining({ detail: { id: 'wifi', checked: true } }));
    expect(spy).toHaveBeenNthCalledWith(2, expect.objectContaining({ detail: { id: 'bt', checked: true } }));
    expect(spy).toHaveBeenNthCalledWith(3, expect.objectContaining({ detail: { id: 'gps', checked: true } }));
  });

  it('toggles each switchesArray item twice and verifies emitted states', async () => {
    const page = await createWithSwitchesArray([
      { id: 'wifi', label: 'Wi-Fi', checked: false },
      { id: 'bt', label: 'Bluetooth', checked: true },
    ]);

    const spy = jest.fn();
    page.win.addEventListener('checkedChanged', spy);

    const inputs = page.root.querySelectorAll('input[type="checkbox"]');

    inputs[0].dispatchEvent(new Event('change'));
    inputs[1].dispatchEvent(new Event('change'));
    await page.waitForChanges();

    inputs[0].dispatchEvent(new Event('change'));
    inputs[1].dispatchEvent(new Event('change'));
    await page.waitForChanges();

    expect(spy).toHaveBeenCalledTimes(4);
    expect(spy).toHaveBeenNthCalledWith(1, expect.objectContaining({ detail: { id: 'wifi', checked: true } }));
    expect(spy).toHaveBeenNthCalledWith(2, expect.objectContaining({ detail: { id: 'bt', checked: false } }));
    expect(spy).toHaveBeenNthCalledWith(3, expect.objectContaining({ detail: { id: 'wifi', checked: false } }));
    expect(spy).toHaveBeenNthCalledWith(4, expect.objectContaining({ detail: { id: 'bt', checked: true } }));
  });

  it('renders no switches when switchesArray is empty', async () => {
    const page = await createWithSwitchesArray([]);
    const inputs = page.root.querySelectorAll('input[type="checkbox"]');
    expect(inputs.length).toBe(0);
    expect(page.root).toMatchSnapshot();
  });

  it('does not emit event when switch is disabled', async () => {
    const page = await createWithSwitchesArray([{ id: 'wifi', label: 'Wi-Fi', checked: false, disabled: true }]);

    const spy = jest.fn();
    page.win.addEventListener('checkedChanged', spy);

    const input = page.root.querySelector('input[type="checkbox"]');
    input.dispatchEvent(new Event('change'));
    await page.waitForChanges();

    expect(spy).not.toHaveBeenCalled();
  });
});
