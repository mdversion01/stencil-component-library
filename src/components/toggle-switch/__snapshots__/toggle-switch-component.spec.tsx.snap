// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toggle-switch-component renders a toggle with toggleTxt and newToggleTxt 1`] = `
<toggle-switch-component checked="true" input-id="txt-toggle" label-txt="Power" new-toggle-txt="{ &quot;on&quot;: &quot;Enabled&quot;, &quot;off&quot;: &quot;Disabled&quot; }" toggle-txt="true">
  <div class="form-group toggle-switch">
    <div class="form-check form-switch">
      <input aria-checked="true" aria-disabled="false" checked="" class="form-check-input" id="txt-toggle" role="switch" type="checkbox" value="">
      <label class="form-check-label" htmlfor="txt-toggle">
        Power
        <span class="toggleTxt-bold">
          On
        </span>
      </label>
    </div>
  </div>
</toggle-switch-component>
`;

exports[`toggle-switch-component renders multiple switches from switchesArray 1`] = `
<toggle-switch-component switches="true">
  <div class="form-group ts-group">
    <div id="toggle-group" role="group">
      <div class="form-check form-switch">
        <input aria-checked="false" aria-disabled="undefined" class="form-check-input" id="toggle-group_option_one" role="switch" type="checkbox">
        <label class="form-check-label" htmlfor="toggle-group_option_one">
          First
        </label>
      </div>
      <div class="form-check form-switch">
        <input aria-checked="true" aria-disabled="undefined" checked="" class="form-check-input" id="toggle-group_option_two" role="switch" type="checkbox">
        <label class="form-check-label" htmlfor="toggle-group_option_two">
          Second
        </label>
      </div>
    </div>
  </div>
</toggle-switch-component>
`;

exports[`toggle-switch-component renders no switches when switchesArray is empty 1`] = `
<toggle-switch-component switches="true">
  <div class="form-group ts-group">
    <div id="toggle-group" role="group"></div>
  </div>
</toggle-switch-component>
`;
